torch/csrc/jit/script/object.h

```diff
@@ -28,34 +28,60 @@ struct TORCH_API Object {
   }
 
   void setattr(const std::string& name, c10::IValue v) {
-    size_t slot = _ivalue()->type()->getAttributeSlot(name);
-    const c10::TypePtr& expected = _ivalue()->type()->getAttribute(slot);
-    TORCH_CHECK(expected, "Module has no attribute '", name, "'");
-    TORCH_CHECK(
-        v.type()->isSubtypeOf(expected),
-        "Expected a value of type '",
-        expected->python_str(),
-        "' for field '",
-        name,
-        "', but found '",
-        v.type()->python_str(),
-        "'");
-    _ivalue()->setSlot(slot, std::move(v));
+    if (auto slot = _ivalue()->type()->findAttributeSlot(name)) {
+      const c10::TypePtr& expected = _ivalue()->type()->getAttribute(*slot);
+      TORCH_CHECK(
+          v.type()->isSubtypeOf(expected),
+          "Expected a value of type '",
+          expected->python_str(),
+          "' for field '",
+          name,
+          "', but found '",
+          v.type()->python_str(),
+          "'");
+      _ivalue()->setSlot(*slot, std::move(v));
+    } else if (auto v = _ivalue()->type()->getConstant(name)) {
+      TORCH_CHECK(
+          false,
+          "Can't set constant '",
+          name,
+          "' which has value:",
+          v);
+    } else {
+      TORCH_CHECK(
+          false,
+          "Module has no attribute '", name, "'");
+    }
   }
 
   c10::IValue attr(const std::string& name) const {
-    return _ivalue()->getAttr(name);
+    if (auto r = _ivalue()->type()->findAttributeSlot(name)) {
+      return _ivalue()->getSlot(*r);
+    }
+    if (auto v = _ivalue()->type()->getConstant(name)) {
```

Comment:
Why is it called `findAttribute...` but `getConstant`? These names feel inconsistent and surprising here.

Reply:
first one is finding the slot and get the value from object, but second one is getting constant directly from type, I think it's reasonable to be different since we do get ivalue from different places. I can change this to a check as well though